README.md
# Razer Drivers

**This is a complete rewrite and cleanup based on the [razer-drivers-legacy](https://github.com/openrazer/razer-drivers-legacy) project. If you are missing a specific device support, don't hesitate to contact me.**

This repository holds the source code for the linux kernel drivers to support various Razer devices.
All notable changes to this project will be documented in the **[CHANGELOG.md](CHANGELOG.md)** file.


## Supported Devices

| NAME                     | TYPE     | DRIVER SUPPORT                                                                            |
|:-------------------------|:---------|:------------------------------------------------------------------------------------------|
| Razer Blade Stealth 2016 | Keyboard | brightness control, lid logo control, FN mode control, all lighting modes, set key colors |
| Razer Blade 14 2016      | Keyboard | brightness control, lid logo control, FN mode control, all lighting modes, set key colors |
| BlackWidow Chroma        | Keyboard | brightness control, macro keys, all lighting modes, set key colors                        |


## Installation

### Packages

- [ArchLinux AUR Package](https://aur.archlinux.org/packages/razer-drivers-dkms/)

### Automatic installation with DKMS

This requires DKMS to be installed.

```
    make DESTDIR=/ install_dkms
```

### Manual installation

This requires recompilation on every kernel update.

```
    make all
    make DESTDIR=/ install
```


## TODO

- Create a debian package.
- BlackWidow Chroma: add support for game mode
- Add information from https://github.com/pez2001/razer_chroma_drivers/wiki


## Authors

A list of all authors contributed to this project can be found in the **[AUTHORS.md](AUTHORS.md)** file.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This project follows the [Semantic Versioning](http://semver.org/).

## v1.0.0 - 2016-07-25
- Initial first release.


scripts/test.sh
#!/bin/bash
# This script tests all available driver functionalities.

# Exit on error.
set -e

DRIVER_PATH="/sys/bus/hid/drivers/hid-razer"
DEVICE_PATH=""
declare -a DEVICES

COLOR='\033[0;31m'
NC='\033[0m' # No Color


if [[ ! -d "$DRIVER_PATH" ]]; then
    echo "driver is not loaded."
    exit 1
fi

# Find all available supported devices.
for d in "$DRIVER_PATH"/*
do
    if [[ $d == *:*:*\.* ]]; then
        DEVICES+=($(basename "$d"))
    fi
done

if [[ "${#DEVICES[@]}" -le "0" ]]; then
    echo "no supported Razer device attached."
    exit 1
elif [[ "${#DEVICES[@]}" -eq "1" ]]; then
    DEVICE_PATH="$DRIVER_PATH/${DEVICES[0]}"
else
    echo "Available devices:"

    i=1
    for d in ${DEVICES[@]}; do
        echo -n " $i) "
        cat "$DRIVER_PATH/$d/device_type" | tr -d '\n'
        echo " ($d)"

        i=$[$i +1]
    done

    read -p "Choose device [1]: " d
    d=${d:-1}
    d=$[$d -1]
    d="${DEVICES[$d]}"

    if [[ "$d" == "" ]]; then
        echo "invalid device path."
        exit 1
    fi

    DEVICE_PATH="$DRIVER_PATH/$d"
fi


echo "Using Razer device $DEVICE_PATH"

echo -n -e "${COLOR}get_firmware_version:${NC} "
cat "$DEVICE_PATH/get_firmware_version"

echo -n -e "${COLOR}get_serial:${NC} "
cat "$DEVICE_PATH/get_serial"

echo -n -e "${COLOR}device_type:${NC} "
cat "$DEVICE_PATH/device_type"

echo -e "${COLOR}get_info:${NC}"
cat "$DEVICE_PATH/get_info"

echo -n -e "${COLOR}brightness:${NC} "
cat "$DEVICE_PATH/brightness"

echo -e "${COLOR}set brightness to${NC} 50"
echo -n "50" > "$DEVICE_PATH/brightness"

echo -n -e "${COLOR}brightness:${NC} "
cat "$DEVICE_PATH/brightness"

sleep 2

echo -e "${COLOR}set brightness to${NC} 0"
echo -n "0" > "$DEVICE_PATH/brightness"

echo -n -e "${COLOR}brightness:${NC} "
cat "$DEVICE_PATH/brightness"

sleep 2

echo -e "${COLOR}set brightness to${NC} 255"
echo -n "255" > "$DEVICE_PATH/brightness"

echo -n -e "${COLOR}brightness:${NC} "
cat "$DEVICE_PATH/brightness"

sleep 2


# set_logo
if [[ -e "$DEVICE_PATH/set_logo" ]]; then
    echo -e "${COLOR}set logo${NC} off"
    echo -n "0" > "$DEVICE_PATH/set_logo"
    sleep 2
    echo -e "${COLOR}set logo${NC} on"
    echo -n "1" > "$DEVICE_PATH/set_logo"
    sleep 2
fi

# set_fn_toggle
if [[ -e "$DEVICE_PATH/set_fn_toggle" ]]; then
    echo -e "${COLOR}set fn mode${NC} 1"
    echo -n "1" > "$DEVICE_PATH/set_fn_toggle"
    sleep 2
    echo -e "${COLOR}set fn mode${NC} 0"
    echo -n "0" > "$DEVICE_PATH/set_fn_toggle"
    sleep 2
fi

# mode_none
if [[ -e "$DEVICE_PATH/mode_none" ]]; then
    echo -e "${COLOR}set mode: ${NC} none"
    echo -n "1" > "$DEVICE_PATH/mode_none"
    sleep 2
fi

# mode_static
if [[ -e "$DEVICE_PATH/mode_static" ]]; then
    echo -e "${COLOR}set mode: ${NC} static"
    echo -e -n "\xFF\x00\xFF" > "$DEVICE_PATH/mode_static"
    sleep 3
fi

# mode_custom
if [[ -e "$DEVICE_PATH/mode_custom" ]]; then
    echo -e "${COLOR}set mode: ${NC} custom"
    echo -n "1" > "$DEVICE_PATH/mode_custom"
    sleep 3
fi

# mode_wave
if [[ -e "$DEVICE_PATH/mode_wave" ]]; then
    echo -e "${COLOR}set mode: ${NC} wave"
    echo -n "1" > "$DEVICE_PATH/mode_wave"
    sleep 3
    echo -n "2" > "$DEVICE_PATH/mode_wave"
    sleep 3
fi

# mode_spectrum
if [[ -e "$DEVICE_PATH/mode_spectrum" ]]; then
    echo -e "${COLOR}set mode: ${NC} spectrum"
    echo -n "1" > "$DEVICE_PATH/mode_spectrum"
    sleep 5
fi

# mode_reactive
if [[ -e "$DEVICE_PATH/mode_reactive" ]]; then
    echo -e "${COLOR}set mode: ${NC} reactive"
    echo -e -n "\x02\xFF\x00\xFF" > "$DEVICE_PATH/mode_reactive"
    sleep 5
fi

# mode_starlight
if [[ -e "$DEVICE_PATH/mode_starlight" ]]; then
    echo -e "${COLOR}set mode: ${NC} starlight"
    echo -e "random colors"
    echo -e -n "\x03" > "$DEVICE_PATH/mode_starlight"
    sleep 5
    echo -e "one color"
    echo -e -n "\x03\xFF\x00\xFF" > "$DEVICE_PATH/mode_starlight"
    sleep 5
    echo -e "two colors"
    echo -e -n "\x03\xFF\x00\xFF\xFF\xFF\xFF" > "$DEVICE_PATH/mode_starlight"
    sleep 5
fi

# mode_breath
if [[ -e "$DEVICE_PATH/mode_breath" ]]; then
    echo -e "${COLOR}set mode: ${NC} breath"
    echo -e "random colors"
    echo -n "1" > "$DEVICE_PATH/mode_breath"
    sleep 7
    echo -e "one color"
    echo -e -n "\xFF\x00\xFF" > "$DEVICE_PATH/mode_breath"
    sleep 7
    echo -e "two colors"
    echo -e -n "\xFF\xFF\x00\xFF\xFF\xFF" > "$DEVICE_PATH/mode_breath"
    sleep 7
fi


src/hid-razer.h
/*
 * Razer Kernel Drivers
 * Copyright (c) 2016 Roland Singer <roland.singer@desertbit.com>
 * Based on Tim Theede <pez2001@voyagerproject.de> razer_chroma_drivers project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#ifndef __HID_RAZER_H
#define __HID_RAZER_H

#include <linux/types.h>



//#################//
//### Constants ###//
//#################//

// Report indexes.
#define RAZER_DEFAULT_REPORT_INDEX  0x02

// Keyboard rows and columns:
#define RAZER_STEALTH_2016_ROWS     0x06
#define RAZER_STEALTH_2016_COLUMNS  0x10

#define RAZER_BLADE_14_2016_ROWS    0x06
#define RAZER_BLADE_14_2016_COLUMNS 0x10

#define RAZER_BLACKWIDOW_CHROMA_ROWS    0x06
#define RAZER_BLACKWIDOW_CHROMA_COLUMNS 0x16



//#############//
//### Types ###//
//#############//

struct razer_data {
    char macro_keys_state;
    char fn_toggle_state;
};

#endif // __HID_RAZER_H


src/hid-razer-common.c
/*
 * Razer Kernel Drivers
 * Copyright (c) 2016 Roland Singer <roland.singer@desertbit.com>
 * Based on Tim Theede <pez2001@voyagerproject.de> razer_chroma_drivers project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include <linux/kernel.h>
#include <linux/slab.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/usb/input.h>
#include <linux/hid.h>

#include "hid-razer-common.h"

//###########################//
//### Version Information ###//
//###########################//

MODULE_AUTHOR("Roland Singer <roland.singer@desertbit.com>");
MODULE_DESCRIPTION("USB Razer common driver");
MODULE_LICENSE("GPL v2");
MODULE_VERSION("1.0.0");



//##########################//
//### Exported Functions ###//
//##########################//

/*
 * Initialize a razer device struct.
 */
int razer_init_device(struct razer_device *razer_dev,
    struct usb_device *usb_dev)
{
    razer_dev->data     = NULL;
    razer_dev->usb_dev  = usb_dev;
    mutex_init(&razer_dev->lock);

    return 0;
}

EXPORT_SYMBOL_GPL(razer_init_device);



/*
 * Get an initialised razer report
 */
struct razer_report razer_new_report(unsigned char command_class,
    unsigned char command_id, unsigned char data_size)
{
    struct razer_report new_report;
    memset(&new_report, 0, sizeof(struct razer_report));

    new_report.status               = RAZER_STATUS_NEW_COMMAND;
    new_report.transaction_id       = 0xFF;
    new_report.remaining_packets    = 0x00;
    new_report.protocol_type        = 0x00;
    new_report.reserved             = 0x00;
    new_report.command_class        = command_class;
    new_report.command_id           = command_id;
    new_report.data_size            = data_size;

    return new_report;
}

EXPORT_SYMBOL_GPL(razer_new_report);



/*
 * Send an USB control report to the device.
 * Returns 0 on success.
 */
int _razer_send(struct razer_device *razer_dev, struct razer_report* report)
{
    const uint size = sizeof(struct razer_report);
    char *buf;
    int len;

    buf = kmemdup(report, size, GFP_KERNEL);
    if (buf == NULL) {
        return -ENOMEM;
    }

    len = usb_control_msg(razer_dev->usb_dev,
            usb_sndctrlpipe(razer_dev->usb_dev, 0),
            HID_REQ_SET_REPORT,                                 // Request
            USB_TYPE_CLASS | USB_RECIP_INTERFACE | USB_DIR_OUT, // RequestType
            0x300,                                              // Value
            razer_dev->report_index,                            // Index
            buf,                                                // Data
            size,                                               // Length
            USB_CTRL_SET_TIMEOUT);

    usleep_range(600, 800);

    kfree(buf);

    return ((len < 0) ? len : ((len != size) ? -EIO : 0));
}

int razer_send(struct razer_device *razer_dev, struct razer_report* report)
{
    int retval;

    mutex_lock(&razer_dev->lock);
    retval = _razer_send(razer_dev, report);
    mutex_unlock(&razer_dev->lock);

    return retval;
}

EXPORT_SYMBOL_GPL(razer_send);



/*
 * Get a response from the razer device.
 * Returns 0 on success.
 */
int _razer_receive(struct razer_device *razer_dev, struct razer_report* report)
{
    const uint size = sizeof(struct razer_report);
    int len;

    memset(report, 0, size);

    len = usb_control_msg(razer_dev->usb_dev,
            usb_rcvctrlpipe(razer_dev->usb_dev, 0),
            HID_REQ_GET_REPORT,                                   // Request
            USB_TYPE_CLASS | USB_RECIP_INTERFACE | USB_DIR_IN,    // RequestType
            0x300,                                                // Value
            razer_dev->report_index,                              // Index
            report,                                               // Data
            size,
            USB_CTRL_SET_TIMEOUT);

    usleep_range(600, 800);

    return ((len < 0) ? len : ((len != size) ? -EIO : 0));
}

int razer_receive(struct razer_device *razer_dev, struct razer_report* report)
{
    int retval;

    mutex_lock(&razer_dev->lock);
    retval = _razer_receive(razer_dev, report);
    mutex_unlock(&razer_dev->lock);

    return retval;
}

EXPORT_SYMBOL_GPL(razer_receive);



/*
 * Send a report and wait for a response.
 * Returns 0 on success.
 */
int _razer_send_with_response(struct razer_device *razer_dev,
    struct razer_report* request_report, struct razer_report* response_report)
{
    int retval, r;

    retval = _razer_send(razer_dev, request_report);
    if (retval != 0) {
        return retval;
    }

    // Retry 40 times when busy -> 125 milliseconds -> max 5 seconds wait
    for (r=0; r < 40; r++) {
        retval = _razer_receive(razer_dev, response_report);
        if (retval != 0) {
            return retval;
        }

        if (response_report->command_class != request_report->command_class ||
                response_report->command_id != request_report->command_id)
        {
            dev_err(&razer_dev->usb_dev->dev,
                    "razer_send_with_response: "
                    "response commands do not match: "
                    "Request Class: %d "
                    "Request ID: %d "
                    "Response Class: %d "
                    "Response ID: %d\n",
                    request_report->command_class,
                    request_report->command_id,
                    response_report->command_class,
                    response_report->command_id);
            return -EINVAL;
        }

        switch (response_report->status) {
        case RAZER_STATUS_SUCCESS:
            return 0;

        case RAZER_STATUS_BUSY:
            msleep(125);
            continue;

        case RAZER_STATUS_FAILURE:
        case RAZER_STATUS_TIMEOUT:
        case RAZER_STATUS_NOT_SUPPORTED:
            return -EINVAL;

        default:
            dev_err(&razer_dev->usb_dev->dev,
                    "razer_send_with_response: "
                    "unknown response status 0x%x\n",
                    response_report->status);
            return -EINVAL;
        }
    }

    dev_err(&razer_dev->usb_dev->dev,
            "razer_send_with_response: request failed: device is busy\n");

    return -EBUSY;
}

int razer_send_with_response(struct razer_device *razer_dev,
    struct razer_report* request_report, struct razer_report* response_report)
{
    int retval;

    mutex_lock(&razer_dev->lock);
    retval = _razer_send_with_response(razer_dev, request_report, response_report);
    mutex_unlock(&razer_dev->lock);

    return retval;
}

EXPORT_SYMBOL_GPL(razer_send_with_response);



/*
 * Send a report, wait for a response and check the response status.
 * Returns 0 on success.
 */
int razer_send_check_response(struct razer_device *razer_dev,
    struct razer_report* request_report)
{
    struct razer_report response_report;

    return razer_send_with_response(razer_dev, request_report, &response_report);
}

EXPORT_SYMBOL_GPL(razer_send_check_response);



/*
 * Calculate the checksum for the usb message
 *
 * Checksum byte is stored in the 2nd last byte in the messages payload.
 * The checksum is generated by XORing all the bytes in the report starting
 * at byte number 2 (0 based) and ending at byte 88.
 */
unsigned char razer_calculate_crc(struct razer_report *report)
{
    // Second to last byte of report is a simple checksum.
    // Just xor all bytes up with overflow and you are done.
    unsigned char crc = 0;
    unsigned char *_report = (unsigned char*)report;

    unsigned int i;
    for(i = 2; i < 88; i++) {
        crc ^= _report[i];
    }

    return crc;
}

EXPORT_SYMBOL_GPL(razer_calculate_crc);



/*
 * Detailed error print
 */
void razer_print_err_report(struct razer_report* report, char* driver_name, char* message)
{
    printk(KERN_WARNING "%s: %s. Status: %02x Transaction ID: %02x Data Size: %02x Command Class: %02x Command ID: %02x Params: %02x%02x%02x%02x%02x%02x%02x%02x%02x%02x\n",
        driver_name,
        message,
        report->status,
        report->transaction_id,
        report->data_size,
        report->command_class,
        report->command_id,
        report->arguments[0], report->arguments[1], report->arguments[2],
        report->arguments[3], report->arguments[4], report->arguments[5],
        report->arguments[6], report->arguments[7], report->arguments[8],
        report->arguments[9]);
}

EXPORT_SYMBOL_GPL(razer_print_err_report);


src/hid-razer-kbd_driver.c
/*
 * Razer Kernel Drivers
 * Copyright (c) 2016 Roland Singer <roland.singer@desertbit.com>
 * Based on Tim Theede <pez2001@voyagerproject.de> razer_chroma_drivers project.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 */

#include <linux/kernel.h>
#include <linux/slab.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/usb/input.h>
#include <linux/hid.h>
#include <linux/dmi.h>

#include "hid-razer-common.h"



//###########################//
//### Version Information ###//
//###########################//

#define DRIVER_AUTHOR "Roland Singer <roland.singer@desertbit.com>"
#define DRIVER_DESC "USB HID Razer Keyboard"
#define DRIVER_LICENSE "GPL v2"

MODULE_AUTHOR(DRIVER_AUTHOR);
MODULE_DESCRIPTION(DRIVER_DESC);
MODULE_LICENSE(DRIVER_LICENSE);



//#################//
//### Constants ###//
//#################//

#define RAZER_KBD_WAIT_MIN_US 600
#define RAZER_KBD_WAIT_MAX_US 800

// Device IDs
#define USB_DEVICE_ID_RAZER_BLADE_STEALTH_2016 0x0205
#define USB_DEVICE_ID_RAZER_BLADE_14_2016 0x020F



//#############//
//### Types ###//
//#############//

struct razer_kbd_device {
    struct usb_device *usbdev;
    struct hid_device *hiddev;
};



//########################//
//### Helper functions ###//
//########################//

// Send a report to the keyboard.
int razer_set_report(struct usb_device *usb_dev,void const *data)
{
    return razer_send_control_msg(usb_dev, data, 0x02, RAZER_KBD_WAIT_MIN_US, RAZER_KBD_WAIT_MAX_US);
}

// Send a report to the keyboard and obtain a response report.
int razer_get_report(struct usb_device *usb_dev,
    struct razer_report *request_report,
    struct razer_report *response_report)
{
    return razer_get_usb_response(usb_dev, 0x02, request_report, 0x02,
        response_report, RAZER_KBD_WAIT_MIN_US, RAZER_KBD_WAIT_MAX_US);
}

// Reset the keyboard.
int razer_reset(struct usb_device *usb_dev)
{
    int retval;
    struct razer_report report = new_razer_report(0x03, 0x00, 0x05);
    report.arguments[0] = 0x01; // LED Class
    report.arguments[1] = 0x08; // LED ID Game mode
    report.arguments[2] = 0x00; // Off
    report.crc = razer_calculate_crc(&report);
    retval = razer_set_report(usb_dev, &report);
    return retval;
}

// Get the firmware version.
int razer_get_firmware_version(struct usb_device *usb_dev, unsigned char* fw_string)
{
    int retval;
    struct razer_report response_report;
    struct razer_report request_report = new_razer_report(0x00, 0x81, 0x02);
    request_report.crc = razer_calculate_crc(&request_report);

    retval = razer_get_report(usb_dev, &request_report, &response_report);
    if(retval != 0)
    {
        razer_print_erroneous_report(&response_report, "hid-razer-kbd", "Invalid Report Length");
        return -1;
    }

    if(response_report.status == 0x02 &&
        response_report.command_class == 0x00 &&
        response_report.command_id.id == 0x81)
    {
        sprintf(fw_string, "v%d.%d", response_report.arguments[0], response_report.arguments[1]);
        retval = response_report.arguments[2];
    }
    else {
        razer_print_erroneous_report(&response_report, "hid-razer-kbd", "Invalid Report Type");
        return -1;
    }

    return retval;
}

// Get brightness of the keyboard.
int razer_get_brightness(struct usb_device *usb_dev)
{
    int retval;
    struct razer_report response_report;

    // Class LED Lighting, Command Set state, 3 Bytes of parameters
    struct razer_report request_report = new_razer_report(0x03, 0x83, 0x03);
    request_report.arguments[0] = 0x01;
    request_report.data_size = 0x02;
    request_report.command_class = 0x0E;
    request_report.command_id.id = 0x84;

    // Calculate the checksum
    request_report.crc = razer_calculate_crc(&request_report);

    retval = razer_get_report(usb_dev, &request_report, &response_report);
    if(retval != 0) {
        razer_print_erroneous_report(&response_report, "hid-razer-kbd", "Invalid Report Length");
        return -1;
    }

    // For the Razer Blades
    if(response_report.status == 0x02 &&
        response_report.command_class == 0x0E &&
        response_report.command_id.id == 0x84)
    {
        retval = response_report.arguments[1];
    }
    else {
        razer_print_erroneous_report(&response_report, "hid-razer-kbd", "Invalid Report Type");
        return -1;
    }

    return retval;
}

// Set the keyboard brightness.
int razer_set_brightness(struct usb_device *usb_dev, unsigned char brightness)
{
    int retval;

    struct razer_report report = new_razer_report(0x03, 0x03, 0x03);
    report.arguments[0] = 0x01;
    report.data_size = 0x02;
    report.command_class = 0x0E;
    report.command_id.id = 0x04;
    report.arguments[1] = brightness;

    // Calculate the checksum
    report.crc = razer_calculate_crc(&report);

    retval = razer_set_report(usb_dev, &report);
    if(retval != 0) {
        razer_print_erroneous_report(&report, "hid-razer-kbd", "request failed");
        return -1;
    }

    return 0;
}

// Set the logo lighting state (on/off only)
int razer_set_logo(struct usb_device *usb_dev, unsigned char state)
{
    int retval;
    struct razer_report report = new_razer_report(0x03, 0x00, 0x03);

    if (state != 0 && state != 1) {
        printk(KERN_WARNING "hid-razer-kbd: logo lighting state must be either 0 or 1. Got: %d", state);
        return -1;
    }

    report.arguments[0] = 0x01; // LED Class
    report.arguments[1] = 0x04; // LED ID, Logo
    report.arguments[2] = state; // State
    report.crc = razer_calculate_crc(&report);

    retval = razer_set_report(usb_dev, &report);
    if(retval != 0) {
        razer_print_erroneous_report(&report, "hid-razer-kbd", "request failed");
        return -1;
    }

    return 0;
}

// Toggle FN key
int razer_set_fn_toggle(struct usb_device *usb_dev, unsigned char state)
{
    int retval;
    struct razer_report report = new_razer_report(0x02, 0x06, 0x02);

    if (state != 0 && state != 1) {
        printk(KERN_WARNING "hid-razer-kbd: Toggle FN, state must be either 0 or 1. Got: %d", state);
        return -1;
    }

    report.arguments[0] = 0x00;
    report.arguments[1] = state; // State
    report.crc = razer_calculate_crc(&report);

    retval = razer_set_report(usb_dev, &report);
    if(retval != 0) {
        razer_print_erroneous_report(&report, "hid-razer-kbd", "request failed");
        return -1;
    }

    return 0;
}



//#########################//
//### Device Attributes ###//
//#########################//

/*
 * Read device file "get_serial"
 * Gets the serial number from the device,
 * Returns a string.
 */
static ssize_t razer_attr_read_get_serial(struct device *dev,
    struct device_attribute *attr, char *buf)
{
    char serial_string[100] = "";

    // Stealth and the Blade does not have serial via USB, so get it from the DMI table.
    strcpy(serial_string, dmi_get_system_info(DMI_PRODUCT_SERIAL));
    return sprintf(buf, "%s\n", &serial_string[0]);
}



/**
 * Read device file "get_firmware_version"
 * Gets the firmware version from the device,
 * Returns a string.
 */
static ssize_t razer_attr_read_get_firmware_version(struct device *dev,
    struct device_attribute *attr, char *buf)
{
    char fw_string[100] = "";
    struct usb_interface *intf = to_usb_interface(dev->parent);
    struct usb_device *usb_dev = interface_to_usbdev(intf);

    razer_get_firmware_version(usb_dev, &fw_string[0]);
    return sprintf(buf, "%s\n", &fw_string[0]);
}



/**
 * Read device file "device_type"
 * Returns a friendly device type string.
 */
static ssize_t razer_attr_read_device_type(struct device *dev,
    struct device_attribute *attr, char *buf)
{
    struct usb_interface *intf = to_usb_interface(dev->parent);
    struct usb_device *usb_dev = interface_to_usbdev(intf);

    char *device_type;

    switch (usb_dev->descriptor.idProduct)
    {
        case USB_DEVICE_ID_RAZER_BLADE_STEALTH_2016:
            device_type = "Razer Blade Stealth 2016\n";
            break;

        case USB_DEVICE_ID_RAZER_BLADE_14_2016:
            device_type = "Razer Blade 14 2016\n";
            break;

        default:
            device_type = "Unknown Device\n";
    }

    return sprintf(buf, device_type);
}



/*
 * Read device file "brightness"
 * Returns the brightness value from 0-255.
 */
static ssize_t razer_attr_read_brightness(struct device *dev,
    struct device_attribute *attr, char *buf)
{
    struct usb_interface *intf = to_usb_interface(dev->parent);
    struct usb_device *usb_dev = interface_to_usbdev(intf);

    int brightness = razer_get_brightness(usb_dev);
    return sprintf(buf, "%d\n", brightness);
}



/*
 * Write device file "rightness"
 * Sets the brightness to the ASCII number written to this file. Values from 0-255
 */
static ssize_t razer_attr_write_brightness(struct device *dev,
    struct device_attribute *attr, const char *buf, size_t count)
{
    struct usb_interface *intf = to_usb_interface(dev->parent);
    struct usb_device *usb_dev = interface_to_usbdev(intf);
    int temp = simple_strtoul(buf, NULL, 10);
    razer_set_brightness(usb_dev, (unsigned char)temp);
    return count;
}



/**
 * Write device file "set_logo"
 * Sets the logo lighting state to the ASCII number written to this file.
 */
static ssize_t razer_attr_write_set_logo(struct device *dev,
    struct device_attribute *attr, const char *buf, size_t count)
{
    struct usb_interface *intf = to_usb_interface(dev->parent);
    struct usb_device *usb_dev = interface_to_usbdev(intf);
    int state = simple_strtoul(buf, NULL, 10);
    razer_set_logo(usb_dev, (unsigned char)state);
    return count;
}



/*
 * Write device file "set_fn_toggle"
 * Sets the logo lighting state to the ASCII number written to this file.
 * If 0 should mean that the F-keys work as normal F-keys
 * If 1 should mean that the F-keys act as if the FN key is held
 */
static ssize_t razer_attr_write_set_fn_toggle(struct device *dev,
    struct device_attribute *attr, const char *buf, size_t count)
{
    struct usb_interface *intf = to_usb_interface(dev->parent);
    struct usb_device *usb_dev = interface_to_usbdev(intf);
    int state = simple_strtoul(buf, NULL, 10);
    razer_set_fn_toggle(usb_dev, (unsigned char)state);
    return count;
}



//######################################//
//### Set up the device driver files ###//
//######################################//

/*
 * Read only is 0444
 * Write only is 0220
 * Read and write is 0664
 */

static DEVICE_ATTR(get_serial,              0444, razer_attr_read_get_serial,           NULL);
static DEVICE_ATTR(get_firmware_version,    0444, razer_attr_read_get_firmware_version, NULL);
static DEVICE_ATTR(device_type,             0444, razer_attr_read_device_type,          NULL);
static DEVICE_ATTR(brightness,              0660, razer_attr_read_brightness, razer_attr_write_brightness);

static DEVICE_ATTR(set_logo,        0220, NULL, razer_attr_write_set_logo);
static DEVICE_ATTR(set_fn_toggle,   0220, NULL, razer_attr_write_set_fn_toggle);



//#############################//
//### Driver Main Functions ###//
//#############################//

/*
 * Probe method is ran whenever a device is bound to the driver
 */
static int razer_kbd_probe(struct hid_device *hdev,
             const struct hid_device_id *id)
{
    int retval;
    struct usb_interface *intf = to_usb_interface(hdev->dev.parent);
    struct usb_device *usb_dev = interface_to_usbdev(intf);
    struct razer_kbd_device *dev = NULL;

    dev = kzalloc(sizeof(struct razer_kbd_device), GFP_KERNEL);
    if(!dev) {
        dev_err(&intf->dev, "failed allocating device descriptor\n");
        return -ENOMEM;
    }

    // Default files
    retval = device_create_file(&hdev->dev, &dev_attr_get_serial);
    if (retval)
        goto exit_free;
    retval = device_create_file(&hdev->dev, &dev_attr_get_firmware_version);
    if (retval)
        goto exit_free;
    retval = device_create_file(&hdev->dev, &dev_attr_device_type);
    if (retval)
        goto exit_free;
    retval = device_create_file(&hdev->dev, &dev_attr_brightness);
    if (retval)
        goto exit_free;


    // Custom files depending on the device support.
    if(usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLADE_STEALTH_2016 ||
            usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLADE_14_2016)
    {
        retval = device_create_file(&hdev->dev, &dev_attr_set_logo);
        if (retval)
            goto exit_free;
        retval = device_create_file(&hdev->dev, &dev_attr_set_fn_toggle);
        if (retval)
            goto exit_free;
    }


    hid_set_drvdata(hdev, dev);

    retval = hid_parse(hdev);
    if(retval)    {
        hid_err(hdev, "parse failed\n");
        goto exit_free;
    }
    retval = hid_hw_start(hdev, HID_CONNECT_DEFAULT);
    if (retval) {
        hid_err(hdev, "hw start failed\n");
        goto exit_free;
    }

    razer_reset(usb_dev);
    usb_disable_autosuspend(usb_dev);

    return 0;
exit_free:
    kfree(dev);
    return retval;
}



/*
 * Driver unbind function
 */
static void razer_kbd_disconnect(struct hid_device *hdev)
{
    struct razer_kbd_device *dev;
    struct usb_interface *intf = to_usb_interface(hdev->dev.parent);
    struct usb_device *usb_dev = interface_to_usbdev(intf);

    dev = hid_get_drvdata(hdev);
    if(!dev)    {
        dev_err(&intf->dev, "disconnect failed: failed to obtain device descriptor\n");
        return;
    }

    // Remove the default files
    device_remove_file(&hdev->dev, &dev_attr_get_firmware_version);
    device_remove_file(&hdev->dev, &dev_attr_get_serial);
    device_remove_file(&hdev->dev, &dev_attr_device_type);
    device_remove_file(&hdev->dev, &dev_attr_brightness);


    // Custom files depending on the device support.
    if(usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLADE_STEALTH_2016 ||
            usb_dev->descriptor.idProduct == USB_DEVICE_ID_RAZER_BLADE_14_2016)
    {
        device_remove_file(&hdev->dev, &dev_attr_set_logo);
        device_remove_file(&hdev->dev, &dev_attr_set_fn_toggle);
    }


    hid_hw_stop(hdev);
    kfree(dev);
    dev_info(&intf->dev, "Razer Device disconnected\n");
}



//###############################//
//### Device ID mapping table ###//
//###############################//

static const struct hid_device_id razer_devices[] = {
    { HID_USB_DEVICE(USB_VENDOR_ID_RAZER, USB_DEVICE_ID_RAZER_BLADE_STEALTH_2016) },
    { HID_USB_DEVICE(USB_VENDOR_ID_RAZER, USB_DEVICE_ID_RAZER_BLADE_14_2016) },
    { }
};

MODULE_DEVICE_TABLE(hid, razer_devices);



//############################################//
//### Describes the contents of the driver ###//
//############################################//

static struct hid_driver razer_kbd_driver = {
    .name       = "razer-kbd",
    .id_table   = razer_devices,
    .probe      = razer_kbd_probe,
    .remove     = razer_kbd_disconnect
};

module_hid_driver(razer_kbd_driver);
